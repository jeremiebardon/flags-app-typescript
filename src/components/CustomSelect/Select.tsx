import {
  chakra,
  FormControlOptions,
  HTMLChakraProps,
  omitThemingProps,
  ThemingProps,
  useFormControl,
  useMergeRefs,
  useMultiStyleConfig,
  usePopper,
} from '@chakra-ui/react'
import { mergeWith } from '@chakra-ui/utils'
import * as React from 'react'
import { SelectIcon } from './SelectIcon'
import { useSelect } from 'downshift'

export interface SelectProps
  extends FormControlOptions,
    ThemingProps<'Select'>,
    Omit<
      HTMLChakraProps<'button'>,
      'disabled' | 'required' | 'readOnly' | 'size' | 'value' | 'onChange'
    > {
  placeholder?: string
  value?: string | null | undefined
  onChange?: (item: string | undefined) => void
}

export const CustomSelect = React.forwardRef<HTMLButtonElement, SelectProps>((props, ownRef) => {
  const { id, value, children, placeholder, onChange, ...rest } = omitThemingProps(props)
  const ownButtonProps = useFormControl<HTMLButtonElement>(rest)
  const styles = useMultiStyleConfig('CustomSelect', props)

  const validChildren = React.useMemo(
    () =>
      React.Children.toArray(children)
        .filter<React.ReactElement<{ value: string; children?: React.ReactElement }>>(
          React.isValidElement,
        )
        .filter((child) => 'value' in child.props),
    [children],
  )

  const items = validChildren.map((child) => child.props.value)

  const { isOpen, selectedItem, getToggleButtonProps, getMenuProps, getItemProps } = useSelect({
    id,
    items,
    selectedItem: value,
    onSelectedItemChange: (val: any) => onChange?.(val.selectedItem),
  })

  const { referenceRef: popperRef, getPopperProps } = usePopper({
    enabled: isOpen,
    gutter: 2,
  })
  const { ref: useSelectToggleButtonRef, ...useSelectToggleButtonProps } = getToggleButtonProps()

  const toggleButtonRef = useMergeRefs(ownRef, useSelectToggleButtonRef, popperRef)
  const toggleButtonProps = mergeWith(ownButtonProps, useSelectToggleButtonProps)

  return (
    <chakra.div position="relative">
      <chakra.button
        ref={toggleButtonRef}
        __css={styles.field}
        data-focus-visible-added={isOpen}
        {...toggleButtonProps}
      >
        {validChildren.find((child) => child.props.value === selectedItem)?.props.children ||
          selectedItem || <chakra.span color='gray.300'>{placeholder}</chakra.span>}
        <SelectIcon />
      </chakra.button>
      <chakra.div
        zIndex="1"
        width="100%"
        {...mergeWith(getPopperProps(), {
          style: { visibility: isOpen ? 'visible' : 'hidden' },
        })}
      >
        <chakra.ul __css={styles.menu} data-focus-visible-added={isOpen} {...getMenuProps()}>
          {isOpen &&
            validChildren.map((item, index) =>
              React.cloneElement(item, {
                __css: styles.option,
                ...getItemProps({ item: item.props.value, index }),
              }),
            )}
        </chakra.ul>
      </chakra.div>
    </chakra.div>
  )
})

CustomSelect.displayName = 'CustomSelect';
