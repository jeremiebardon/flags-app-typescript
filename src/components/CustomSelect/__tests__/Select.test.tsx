import { render } from "@testing-library/react";
import { CustomSelect } from "../Select";

describe('<Select />', () => {
  it('should render', () => {
    const wrapper = render(<CustomSelect />);

    expect(wrapper).toMatchSnapshot();
  });
});