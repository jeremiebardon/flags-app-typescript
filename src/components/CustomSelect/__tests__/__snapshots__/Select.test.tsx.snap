// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-79elbk"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="downshift-0-label downshift-0-toggle-button"
          class="css-0"
          data-focus-visible-added="false"
          id="downshift-0-toggle-button"
        >
          <span
            class="css-188y514"
          />
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            role="presentation"
            style="width: 1em; height: 1em; color: currentColor;"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          class="css-1pye1st"
          style="position: absolute; min-width: max-content; inset: 0 auto auto 0; visibility: hidden;"
        >
          <ul
            aria-labelledby="downshift-0-label"
            class="css-0"
            data-focus-visible-added="false"
            id="downshift-0-menu"
            role="listbox"
            tabindex="-1"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-79elbk"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="downshift-0-label downshift-0-toggle-button"
        class="css-0"
        data-focus-visible-added="false"
        id="downshift-0-toggle-button"
      >
        <span
          class="css-188y514"
        />
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          role="presentation"
          style="width: 1em; height: 1em; color: currentColor;"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="css-1pye1st"
        style="position: absolute; min-width: max-content; inset: 0 auto auto 0; visibility: hidden;"
      >
        <ul
          aria-labelledby="downshift-0-label"
          class="css-0"
          data-focus-visible-added="false"
          id="downshift-0-menu"
          role="listbox"
          tabindex="-1"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
